File: src/sorts/InsertionSort.java
package sorts;

/**
 * Created by Chris Keyser on 3/17/2016.
 * Insertion sort implementation
 */
public class InsertionSort extends Sort {
    public InsertionSort(int[] arr) {
        this.arr = arr;
    }

    @Override
    public int[] sort() {
        for(int i = 0; i < arr.length; i++) {
            for(int j = i; j > 0 && arr[j] < arr[j - 1]; j--) {
                swap(j, j - 1);
            }
        }

        return arr;
    }
}

File: go/src/sorts/insertionsort.go
package sorts

func InsertionSort(arr []int) []int {
  sorted := make([]int, len(arr))
  copy(sorted, arr)

  for i := 1; i < len(sorted); i++ {
    for j := i - 1; j > 0 && sorted[i] < sorted[j]; j-- {
      sorted[i], sorted[j] = sorted[j], sorted[i]
    }
  }

  return sorted
}
